crate_universe/private/local_crate_mirror.bzl

```
@@ -0,0 +1,63 @@
+"""`local_crate_mirror` rule implementation.
+
+This is a private implementation detail of crate_universe, and should not be relied on in manually written code.
+
+This is effectively a `local_repository` rule impementation, but where the BUILD.bazel file is generated using the `cargo-bazel render` command.
+"""
+
+load("//crate_universe/private:common_utils.bzl", "execute")
+load("//crate_universe/private:generate_utils.bzl", "get_generator")
+load("//crate_universe/private:urls.bzl", "CARGO_BAZEL_SHA256S", "CARGO_BAZEL_URLS")
+load("//rust/platform:triple.bzl", "get_host_triple")
+
+def _local_crate_mirror_impl(repository_ctx):
+    path = repository_ctx.path(repository_ctx.attr.path)
+
+    host_triple = get_host_triple(repository_ctx)
+
+    generator, _generator_sha256 = get_generator(repository_ctx, host_triple.str)
+
+    execute(repository_ctx, ["bash", "-c", "cp -r {}/* {}/".format(path, repository_ctx.path("."))])
+
+    paths_to_track = execute(repository_ctx, ["bash", "-c", "find {} -type f".format(path)]).stdout.strip().split("\n")
+    for path_to_track in paths_to_track:
+        if path_to_track:
+            repository_ctx.read(path_to_track)
+
+    execute(repository_ctx, [generator, "render", "--options-json", repository_ctx.attr.options_json, "--output-path", repository_ctx.path("BUILD.bazel")])
+
+    repository_ctx.file("WORKSPACE", "")
+
+local_crate_mirror = repository_rule(
```

nit: description about what this repository rule is for?
```suggestion
local_crate_mirror = repository_rule(
    doc = "TODO",
```

---

edit: Oh, I see there's some details in the header. I would either move it or at least say
```suggestion
local_crate_mirror = repository_rule(
    doc = "See `.bzl` header.",
```